include/gzip/decompress.hpp

```diff
@@ -56,7 +56,7 @@ class Decompressor
         // https://github.com/kaorimatz/libdeflate-ruby/blob/0e33da96cdaad3162f03ec924b25b2f4f2847538/ext/libdeflate/libdeflate_ext.c#L340
         // https://github.com/ebiggers/libdeflate/commit/5a9d25a8922e2d74618fba96e56db4fe145510f4
         std::size_t actual_size;
-        std::size_t uncompressed_size_guess = size * 3;
+        std::size_t uncompressed_size_guess = size * 4;
         output.resize(uncompressed_size_guess);
```

Comment:
This probably needs to be re-written to avoid needing to guess the size up front. @artemp could you tackle that?

Reply:
`std::size_t uncompressed_size_guess = size * 4;` - good starting point but it definitely needs a way to dynamically increase capacity if `size * 4` is not large enough.  Approach from the ruby link looks reasonable, here is my patch : 

```patch
diff --git a/include/gzip/decompress.hpp b/include/gzip/decompress.hpp
index badd315..627e670 100644
--- a/include/gzip/decompress.hpp
+++ b/include/gzip/decompress.hpp
@@ -57,17 +57,23 @@ class Decompressor
         // https://github.com/ebiggers/libdeflate/commit/5a9d25a8922e2d74618fba96e56db4fe145510f4
         std::size_t actual_size;
         std::size_t uncompressed_size_guess = size * 4;
-        output.resize(uncompressed_size_guess);
-        enum libdeflate_result result = libdeflate_gzip_decompress(decompressor_,
-                                                                   data,
-                                                                   size,
-                                                                   static_cast<void*>(&output[0]),
-                                                                   uncompressed_size_guess, &actual_size);
-        if (result == LIBDEFLATE_INSUFFICIENT_SPACE)
+        output.reserve(uncompressed_size_guess);
+        enum libdeflate_result result;
+        for (;;)
         {
-            throw std::runtime_error("no space: did not succeed");
+            result = libdeflate_gzip_decompress(decompressor_,
+                                                data,
+                                                size,
+                                                const_cast<char*>(output.data()),
+                                                output.capacity(), &actual_size);
+            if (result != LIBDEFLATE_INSUFFICIENT_SPACE)
+            {
+                break;
+            }
+            output.reserve((output.capacity() << 1) - output.size());
         }
-        else if (result == LIBDEFLATE_SHORT_OUTPUT)
+
+        if (result == LIBDEFLATE_SHORT_OUTPUT)
         {
             throw std::runtime_error("short output: did not succeed");
         }


```

Couple notes: 

* Use `std::vector::reserve()` to increase capacity
* Use `std::vector::data()` to access a pointer to the first element (c++11) which works for empty containers, while `&vec[0]` is not safe and need an extra check.

Reply:
@artemp great fix - can you please commit directly to this branch? But two concerns to keep in mind:

 - Can you think of a situation where `result` will never become == to `LIBDEFLATE_INSUFFICIENT_SPACE` and therefore the loop will run infinitely? Can we protect against that by checking for other error states inside the loop?
 - I wonder if we should avoid calling `output.reserve` because sometimes we may overeserve and the extra cost of over-reserving might overrule the benefit of reserving when it is relatively correct.  This concern of mine comes from seeing how `reserve` is expensive and when wrong hurts perf over at https://github.com/mapbox/carmen-cache/pull/126

Reply:
@springmeyer - 

> Can you think of a situation where result will never become == to LIBDEFLATE_INSUFFICIENT_SPACE and therefore the loop will run infinitely? Can we protect against that by checking for other error states inside the loop?

Good point, I'll add conditional exception in the  loop if allocation size is getting to big.

> I  wonder if we should avoid calling output.reserve because sometimes we may overeserve and the extra cost of over-reserving might overrule the benefit of reserving when it is relatively correct. This concern of mine comes from seeing how reserve is expensive and when wrong hurts perf over at mapbox/carmen-cache#126

Not sure I understand here. Do you mean not to call `reserve` with `uncompressed_size_guess`  before entering the loop? I think we should as it's a good assumption that uncompressed buffer will be larger then compressed one. Perhaps multiple of 3 is a better guess.  Also, inside the loop we need to grow output buffer in case the guessed size is not sufficient. Please, elaborate if I'm missing something. `reserve` is faster in most cases than `resize` 

I'm going to merge this PR+modification and continue work on the branch.

Reply:
> I'm going to push updates and continue work on this branch

👍  I think you'll likely need an easy way to test larger, varied files and a CLI would allow that. So could you finish https://github.com/mapbox/gzip-hpp/pull/29 to aid your testing of this branch?